### PR TITLE
Implement the module [Wallet], according spv's interfaces.

### PR DESCRIPTION
**Pull request by zhaowanli2**

### CODE REVIEW

File: Sources/Elastos/Blockchain/Wallet/CWalletEnviroment.cpp [hunk 1]
```diff
@@ -1,27 +1,59 @@
...
+
+        try {
+            spvMasterWalletMgr = Elastos::ElaWallet::Enviroment::GetMasterWalletManager();
+        }
+        catch (std::logic_error error) {
+            return E_INVALID_ARGUMENT;
+        }
+
+        mMasterWalletManager = new MasterWalletManager(spvMasterWalletMgr);
+        if (mMasterWalletManager == NULL) return E_OUT_OF_MEMORY;
```
Comment (zuohuahua) [inline Sources/Elastos/Blockchain/Wallet/CWalletEnviroment.cpp]: delete spvMasterWalletMgr

File: Sources/Elastos/Blockchain/Wallet/MasterWallet.cpp [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+    std::vector<Elastos::ElaWallet::ISubWallet *> array = mSpvMasterWallet->GetAllSubWallets();
+    const Int32 length = array.size();
+    if (length < 1) {
+        return NOERROR;
+    }
+
+    AutoPtr<ArrayOf<ISubWallet*> > _subWallets = ArrayOf<ISubWallet*>::Alloc(length);
+    for (Int32 i = 0; i < length; i++) {
+        AutoPtr<ISubWallet> subWallet = new SubWallet(array[i]);
+        if (subWallet == NULL) return E_OUT_OF_MEMORY;
```
Comment (zuohuahua) [inline Sources/Elastos/Blockchain/Wallet/MasterWallet.cpp]: free SubWallet in array

File: Sources/Elastos/Blockchain/Wallet/MasterWallet.cpp [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+    /* [in] */ const String& payPassword,
+    /* [in] */ Boolean singleAddress,
+    /* [in] */ Int64 feePerKb,
+    /* [out] */ ISubWallet** subWallet)
+{
+    VALIDATE_NOT_NULL(subWallet);
+    Elastos::ElaWallet::ISubWallet* _subWallet = mSpvMasterWallet->CreateSubWallet(chainID.string(), payPassword.string(),
+                singleAddress, feePerKb);
+    AutoPtr<ISubWallet> wallet = new SubWallet(_subWallet);
+    if (wallet == NULL) return E_OUT_OF_MEMORY;
```
Comment (zuohuahua) [inline Sources/Elastos/Blockchain/Wallet/MasterWallet.cpp]: delete _subWallet

File: Sources/Elastos/Blockchain/Wallet/MasterWallet.cpp [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+    /* [in] */ Boolean singleAddress,
+    /* [in] */ Int32 limitGap,
+    /* [in] */ Int64 feePerKb,
+    /* [out] */ ISubWallet** subWallet)
+{
+    VALIDATE_NOT_NULL(subWallet);
+    Elastos::ElaWallet::ISubWallet* _subWallet = mSpvMasterWallet->RecoverSubWallet(chainID.string(), payPassword.string()
+                    , singleAddress, limitGap, feePerKb);
+    AutoPtr<ISubWallet> wallet = new SubWallet(_subWallet);
+    if (wallet == NULL) return E_OUT_OF_MEMORY;
```
Comment (zuohuahua) [inline Sources/Elastos/Blockchain/Wallet/MasterWallet.cpp]: delete _subWallet

File: Sources/Elastos/Blockchain/Wallet/MasterWalletManager.cpp [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+ECode MasterWalletManager::CreateMasterWallet(
+    /* [in] */ const String& masterWalletId,
+    /* [in] */ const String& language,
+    /* [out] */ IMasterWallet** masterWallet)
+{
+    VALIDATE_NOT_NULL(masterWallet);
+    VALIDATE_NOT_NULL(mSpvMasterWalletMgr);
+    Elastos::ElaWallet::IMasterWallet* spvMasterWallet = mSpvMasterWalletMgr->CreateMasterWallet(masterWalletId.string(), language.string());
+    AutoPtr<IMasterWallet> wallet = new MasterWallet(spvMasterWallet);
+    if (wallet == NULL) return E_OUT_OF_MEMORY;
```
Comment (zuohuahua) [inline Sources/Elastos/Blockchain/Wallet/MasterWalletManager.cpp]: delete spvMasterWallet

File: Sources/Elastos/Blockchain/Wallet/MasterWalletManager.cpp [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+    std::vector<Elastos::ElaWallet::IMasterWallet *> array = mSpvMasterWalletMgr->GetAllMasterWallets();
+    const Int32 length = array.size();
+    if (length < 1) {
+        return NOERROR;
+    }
+
+    AutoPtr<ArrayOf<IMasterWallet*> > _masterWallets = ArrayOf<IMasterWallet*>::Alloc(length);
+    for (Int32 i = 0; i < length; i++) {
+        AutoPtr<IMasterWallet> masterWallet = new MasterWallet(array[i]);
+        if (masterWallet == NULL) return E_OUT_OF_MEMORY;
```
Comment (zuohuahua) [inline Sources/Elastos/Blockchain/Wallet/MasterWalletManager.cpp]: delete IMasterWallet in array

File: Sources/Elastos/Blockchain/Wallet/MasterWalletManager.cpp [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+    /* [out] */ IMasterWallet** masterWallet)
+{
+    VALIDATE_NOT_NULL(masterWallet);
+    VALIDATE_NOT_NULL(mSpvMasterWalletMgr);
+    *masterWallet = NULL;
+
+    Elastos::ElaWallet::IMasterWallet* spvWallet = mSpvMasterWalletMgr->ImportWalletWithKeystore(masterWalletId.string(), keystorePath.string()
+            , backupPassword.string(), payPassword.string(), phrasePassword.string());
+    *masterWallet = new MasterWallet(spvWallet);
+    if ((*masterWallet) == NULL) return E_OUT_OF_MEMORY;
```
Comment (zuohuahua) [inline Sources/Elastos/Blockchain/Wallet/MasterWalletManager.cpp]: delete spvWallet

File: Sources/Elastos/Blockchain/Wallet/MasterWalletManager.cpp [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+{
+    VALIDATE_NOT_NULL(masterWallet);
+    VALIDATE_NOT_NULL(mSpvMasterWalletMgr);
+    *masterWallet = NULL;
+
+    Elastos::ElaWallet::IMasterWallet* spvWallet = mSpvMasterWalletMgr->ImportWalletWithMnemonic(masterWalletId.string(), mnemonic.string()
+                , phrasePassword.string(), payPassword.string(), language.string());
+
+    *masterWallet = new MasterWallet(spvWallet);
+    if ((*masterWallet) == NULL) return E_OUT_OF_MEMORY;
```
Comment (zuohuahua) [inline Sources/Elastos/Blockchain/Wallet/MasterWalletManager.cpp]: delete spvWallet

File: Sources/Elastos/Blockchain/Wallet/CWalletEnviroment.cpp (hunk 1 above)
Comment (zhaowanli2) [reply Sources/Elastos/Blockchain/Wallet/CWalletEnviroment.cpp]: spvMasterWalletMgr will be delete at MasterWalletManager::~MasterWalletManager(){}

File: Sources/Elastos/Blockchain/Wallet/MasterWallet.cpp [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+{
+    VALIDATE_NOT_NULL(subWallets);
+    *subWallets = NULL;
+    std::vector<Elastos::ElaWallet::ISubWallet *> array = mSpvMasterWallet->GetAllSubWallets();
+    const Int32 length = array.size();
+    if (length < 1) {
+        return NOERROR;
+    }
+
+    AutoPtr<ArrayOf<ISubWallet*> > _subWallets = ArrayOf<ISubWallet*>::Alloc(length);
```
[the remark this answers is not in the document]
Comment (zhaowanli2) [reply Sources/Elastos/Blockchain/Wallet/MasterWallet.cpp]: ok

File: Sources/Elastos/Blockchain/Wallet/MasterWallet.cpp (hunk 3 above)
Comment (zhaowanli2) [reply Sources/Elastos/Blockchain/Wallet/MasterWallet.cpp]: _subWallet will be deleted at SubWallet::~SubWallet(){}

File: Sources/Elastos/Blockchain/Wallet/MasterWallet.cpp (hunk 4 above)
Comment (zhaowanli2) [reply Sources/Elastos/Blockchain/Wallet/MasterWallet.cpp]: _subWallet will be deleted at SubWallet::~SubWallet(){}

File: Sources/Elastos/Blockchain/Wallet/MasterWalletManager.cpp (hunk 5 above)
Comment (zhaowanli2) [reply Sources/Elastos/Blockchain/Wallet/MasterWalletManager.cpp]: spvMasterWallet will be deleted at MasterWallet::~MasterWallet(){}

File: Sources/Elastos/Blockchain/Wallet/MasterWalletManager.cpp (hunk 7 above)
Comment (zhaowanli2) [reply Sources/Elastos/Blockchain/Wallet/MasterWalletManager.cpp]: spvWallet will be deleted at MasterWallet::~MasterWallet(){}

File: Sources/Elastos/Blockchain/Wallet/MasterWalletManager.cpp (hunk 8 above)
Comment (zhaowanli2) [reply Sources/Elastos/Blockchain/Wallet/MasterWalletManager.cpp]: spvWallet will be deleted at MasterWallet::~MasterWallet(){}

File: Sources/Elastos/Blockchain/Wallet/SubWallet.cpp [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+    AddSubWalletCallbackNode(callback);
+
+    return NOERROR;
+}
+
+ECode SubWallet::RemoveCallback (
+    /* [in] */ ISubWalletListener* listener)
+{
+    VALIDATE_NOT_NULL(listener);
+    AutoPtr<SubWalletCallback> callback = new SubWalletCallback(listener);
```
Comment (zuohuahua) [inline Sources/Elastos/Blockchain/Wallet/SubWallet.cpp]: check callback is NULL

File: Sources/Elastos/Blockchain/Wallet/SubWallet.cpp [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+    SubWalletCallbackNode* it = &mCallbacks;
+    it = (SubWalletCallbackNode*)(it->Next());
+    for (; NULL != it; it = (SubWalletCallbackNode*)(it->Next())) {
+        if (it->mListener.Get() == callback) {
+            pthread_mutex_unlock(&mCallbacksLock);
+            return;
+        }
+    }
+
+    SubWalletCallbackNode* listenerNode = new SubWalletCallbackNode;
```
Comment (zuohuahua) [inline Sources/Elastos/Blockchain/Wallet/SubWallet.cpp]: check listenerNode is NULL

**Pull request closed without merging**